### ISSUE TITLE
Convergence of multi-element GAP and optimal hyperparameters

### ISSUE THREAD

**Issue by yonghyuklee** [libAtoms/QUIP#588]
Dear GAP developers,

I'm currently working on the three-element crystalline system and trying to converge the GAP in an iterative refinement scheme. I had some experience in single and two-element systems, but I feel that the convergence of the three-element GAP is much slower. I can understand this since the required size of the training data would scale quadratically to the number of elements, but I obscurely doubt this could be the problem of hyperparameter selection or etc...

My current training set includes 567 energies and 100467 forces. I use the hyperparameters as follows:
`gap_fit at_file=training_set.xyz energy_parameter_name=dft_energy force_parameter_name=dft_forces do_copy_at_file=F sparse_separate_file=F gp_file=gap.xml default_sigma={0.001 0.01 0 0} gap={distance_2b cutoff=5.0 n_sparse=25 covariance_type=ard_se delta=0.372 theta_uniform=1.0 sparse_method=uniform add_species=T :soap cutoff=5.0 l_max=4 n_max=8 covariance_type=dot_product add_species=T delta=0.108 atom_sigma=0.5 zeta=4 sparse_method=cur_points n_sparse=2000}`
Most hyperparameters were selected heuristically or from cross-validations. 

![plot](https://github.com/libAtoms/QUIP/assets/18476654/570c53e0-1f54-48c9-a820-1d91f70b9585)

I still have not very satisfying correlation plots, although the test set used here is new samples in my active learning iteration, so that its size is rather small. The training error is also not that bad but is not satisfying. Could you suggest what I can try?
Some of my hunch:
- The test structures are close to ground structure samples, so the forces are very small. Is my current hyperparameter setting not accurate enough for these low-force data?
- n_sparse may not be large enough. I have tested larger values, such as 5000, and it somewhat improves the result. But it is still not a dramatic improvement, and the force correlation plot still looks bad. This also gives a much heavier GAP, and I don't know if I want this.
- Larger n_max=12 has been tested. It didn't give reliable change.
- Tighter regularization values
- Not enough size of the training set

Many thanks.

**Comment by albapa** [comment]
The scale of your training and validation sets are indeed very different - so the plots appear worse than they actually are.
It's hard to say what's wrong - quite possibly nothing. A more important figure would be a training curve, i.e. what happens as you add more data. If the RMSEs go down, you just need more data.

**Comment by MES-physics** [comment]
Hi,
Can you please clarify again what should happen to RMSEs when more data is added?  I thought I had heard that RMSEs are supposed to go down with more data, not up? We want lower RMSEs, correct?( I’m a student.)

**Comment by jesperbygg** [comment]
You can try adding the 3-body descriptor (angle_3b). I have had good success with three-body GAPs for multi-element alloys (even without SOAP, see https://arxiv.org/abs/2203.08458, https://doi.org/10.1103/PhysRevMaterials.6.083801).

**Comment by albapa** [comment]
> Hi, Can you please clarify again what should happen to RMSEs when more data is added? I thought I had heard that RMSEs are supposed to go down with more data, not up? We want lower RMSEs, correct?( I’m a student.)

If you see the RMSEs going down in your training curve as adding more data (as opposed to levelling out or going up), it generally means you can improve your model further with more data.

**Comment by yonghyuklee** [comment]
Thanks a lot for the suggestions!
Checking the convergence of RMSEs makes sense. I can see the convergence, and the force RMSE seems to be oscillating between 250 and 350 meV/Å. Does it look reasonable within my hyperparameters? I'm curious whether I can still get better values by including more structures.

I can also try to include the 3b descriptor. I've never tried it before. I will update the thread afterwards.

**Comment by yonghyuklee** [comment]
The inclusion of the 3B term improves it a bit, but not drastically as well.

![plot](https://github.com/libAtoms/QUIP/assets/18476654/adfd1387-ecaf-48d8-be3e-1831d9f7b736)

**Comment by gabor1** [comment]
I would be curious to see your train and validation errors with just the 2B potential (everything else staying the same, just remove the SOAP part). Is it possible that your soap is doing very little? Try increasing the delta of the SOAP as well, see if that does anything (to e.g. 1.0). Are these 3D periodic configs? do you have stress data and stress predictions ? sometimes it can give a clue. Can you plot the force validation with smaller dots - I want to see if in the middle of that blue patch, there is any correlation at all...

**Comment by yonghyuklee** [comment]
> I would be curious to see your train and validation errors with just the 2B potential (everything else staying the same, just remove the SOAP part). Is it possible that your soap is doing very little? Try increasing the delta of the SOAP as well, see if that does anything (to e.g. 1.0). Are these 3D periodic configs? do you have stress data and stress predictions ? sometimes it can give a clue. Can you plot the force validation with smaller dots - I want to see if in the middle of that blue patch, there is any correlation at all...

Hi Prof. Csanyi,

Thanks for the suggestions. These are 3D periodic configurations, but they are slabs. This is why I didn't include any stress data/prediction.
Here is the 2B potential result:
![plot](https://github.com/libAtoms/QUIP/assets/18476654/375d7176-182b-435a-8c4b-480f868f47dc)

This is the original plot with smaller dots:
![plot](https://github.com/libAtoms/QUIP/assets/18476654/aade096d-9684-4e9e-8b16-45349adaacad)

Lastly, with an increased delta (1.0) for SOAP:
![plot](https://github.com/libAtoms/QUIP/assets/18476654/34b1ec08-5b81-446d-829e-3e51eba0d9d0)

**Comment by gabor1** [comment]
intriguing. so for your validation, soap is doing very little I'd say. so maybe you are overfitting, the 0.01 force sigma is way smaller than your validation error (and even your training error), so try increasing it to 0.1. for your 2B fit, there are two populations in your training data in the force plot: what kind of configurations are those which have large + or - DFT components but nearly zero GAP components?

**Comment by gabor1** [comment]
I note that your validation force data has a dft force range between -1 and 1. so can you show a zoom on your TRAINING force-force correlation plot (2b+soap) just in the -1 - +1 range? I am curious to see if there are two populations there also.

**Comment by gabor1** [comment]
you mentioned that you have a 3-element system. did you not see anything like this with your earlier two element system? Can you say something about what kinds of configurations are these (maybe in a private email, if you don't want to share your data here)

**Comment by MES-physics** [comment]
Can someone please clarify how to see whether we have overfitting or underfitting?  What is it that must be compared?
Thanks.

**Comment by gabor1** [comment]
unless there are other bugs at play to do with data or other workflow procedures, overfitting is a situation in which you errors on your training data are much lower than in your testing data (like in this issue). Underfitting is really just saying that your regularisation is too strong, and you could fit better (both training and testing) if you weakened your regularisation - and you test this by weakening the regularisation.
